homeassistant/components/modern_forms/__init__.py

```diff
@@ -0,0 +1,180 @@
+        await asyncio.gather(
+            *(
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            )
+        )
+    )
+
+    if unload_ok:
+        del hass.data[DOMAIN][entry.entry_id]
+
+    if not hass.data[DOMAIN]:
+        del hass.data[DOMAIN]
+
+    return unload_ok
+
+
+def modernforms_exception_handler(func):
+    """Decorate Modern Forms calls to handle Modern Forms exceptions.
+
+    A decorator that wraps the passed in function, catches Modern Forms errors,
+    and handles the availability of the device in the data coordinator.
+    """
+
+    async def handler(self, *args, **kwargs):
+        try:
+            await func(self, *args, **kwargs)
+            self.coordinator.update_listeners()
+
+        except ModernFormsConnectionError as error:
+            _LOGGER.error("Error communicating with API: %s", error)
+            self.coordinator.last_update_success = False
+            self.coordinator.update_listeners()
+
+        except ModernFormsError as error:
+            _LOGGER.error("Invalid response from API: %s", error)
+
+    return handler
+
+
+class ModernFormsDataUpdateCoordinator(DataUpdateCoordinator[Device]):
+    """Class to manage fetching Modern Forms data from single endpoint."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        host: str,
+    ) -> None:
+        """Initialize global Modern Forms data updater."""
+        self.modernforms = ModernFormsDevice(
+            host, session=async_get_clientsession(hass)
+        )
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+
+    def update_listeners(self) -> None:
+        """Call update on all listeners."""
+        for update_callback in self._listeners:
+            update_callback()
+
+    async def _async_update_data(self) -> ModernFormsDevice:
+        """Fetch data from Modern Forms."""
+        try:
+            return await self.modernforms.update(
+                full_update=not self.last_update_success
+            )
+        except ModernFormsError as error:
+            raise UpdateFailed(f"Invalid response from API: {error}") from error
+
+
+class ModernFormsEntity(CoordinatorEntity[ModernFormsDataUpdateCoordinator]):
+    """Defines a base Modern Forms entity."""
+
+    def __init__(
+        self,
+        *,
+        entry_id: str,
+        coordinator: ModernFormsDataUpdateCoordinator,
+        name: str,
+        icon: str,
+        enabled_default: bool = True,
+    ) -> None:
+        """Initialize the Modern Forms entity."""
+        super().__init__(coordinator)
+        self._attr_enabled_default = enabled_default
+        self._entry_id = entry_id
+        self._attr_icon = icon
+        self._attr_name = name
+        self._unsub_dispatcher = None
+
+
+class ModernFormsDeviceEntity(ModernFormsEntity):
```

Comment:
Unless you plan to add entities which are not part of a device, you can just merge this class with `ModernFormsEntity`

Reply:
Done